list_objects.go

```diff
@@ -96,20 +112,38 @@ func (l *ObjectLister) worker() {
 		}
 	}
 
-	l.wg.Done()
+	return nil
 }
 
-func (l *ObjectLister) retryListObjects(p, continuation string) (*listBucketResult, error) {
+func (l *ObjectLister) retryListObjects(
+	ctx context.Context, p, continuation string,
+) (*listBucketResult, error) {
 	var err error
 	var res *listBucketResult
+	var timer *time.Timer
 	for i := 0; i < l.c.NTry; i++ {
 		opts := listObjectsOptions{MaxKeys: l.maxKeys, Prefix: p, ContinuationToken: continuation}
 		res, err = listObjects(l.c, l.b, opts)
 		if err == nil {
 			return res, nil
 		}
 
-		time.Sleep(time.Duration(math.Exp2(float64(i))) * 100 * time.Millisecond) // exponential back-off
+		// Exponential back-off, reusing the timer if possible:
+		duration := time.Duration(math.Exp2(float64(i))) * 100 * time.Millisecond
+		if timer == nil {
+			timer = time.NewTimer(duration)
+		} else {
+			timer.Reset(duration)
```

Comment:
This could do with a comment explaining why this is safe to do whenever it exists as https://pkg.go.dev/time#Timer.Reset seems to think it's quite important not to use on a running on undrained timer.